src/librustc/ty/context.rs

```diff
@@ -1769,6 +1753,16 @@ impl<'a, 'tcx> Lift<'tcx> for &'a Const<'a> {
     }
 }
 
+impl<'a, 'tcx> Lift<'tcx> for &'a mir::interpret::Allocation {
+    type Lifted = &'tcx mir::interpret::Allocation;
+    fn lift_to_tcx<'b, 'gcx>(&self, tcx: TyCtxt<'b, 'gcx, 'tcx>) -> Option<Self::Lifted> {
+        if tcx.interners.arena.in_arena(*self as *const _) {
+            return Some(unsafe { mem::transmute(*self) });
+        }
+        Some(tcx.intern_const_alloc(mir::interpret::Allocation::clone(self)))
```

Comment:
OK, so, this seems *valid* but it is also quite different from the standard `Lift` pattern, which basically just *checks* whether the value is already allocated. This seems more like an *intern* sort of pattern. 

I'm a bit wary here, because it seems like an unexpected interaction that will be confusing. I see some comments above that try to explain why this is different @oli-obk from the existing code, but I don't quite follow.

I would feel better if we made `Lift` return `None` and then had a `lift_or_intern(&Allocation) -> &'tcx Allocation` that did something like `allocation.lift_to_tcx().unwrap_or(...)`, but I'm guessing that this `Allocation` is maybe embedded inside of something else, and so that wouldn't necessarily work?

(Has @eddyb seen this?)

Reply:
(The usual pattern of course would just be to intern *all* `Allocation`  pointers)

Reply:
Well... I just realized that we indeed intern all `Allocation`s, we just need to check the right arena: https://github.com/rust-lang/rust/blob/d2986970adae36939d13c79e3af34ff6378fad68/src/librustc/ty/context.rs#L1035

Reply:
We could technically make all of this a NOP by adding an additional `'gcx` lifetime to `ty::Const`, but that would infect everything

Reply:
> could basically just assert 

Heh, that's literally what i already did earlier today. I'll push the changes after i make a rebase

Reply:
jup that worked